src/ucp/wireup/wireup_cm.c

```diff
@@ -267,37 +284,50 @@ ucp_cm_ep_priv_data_pack(ucp_ep_h ep, const ucp_tl_bitmap_t *tl_bitmap,
         goto err;
     }
 
-    cm_idx = ucp_ep_ext_control(ep)->cm_idx;
-    if (worker->cms[cm_idx].attr.max_conn_priv <
-        ucp_cm_priv_data_length(ucp_addr_size)) {
-        log_level = can_fallback ?
-                    UCS_LOG_LEVEL_DIAG : UCS_LOG_LEVEL_ERROR;
+    sa_data_length = ucp_cm_priv_data_length(ucp_addr_size, sa_data_version);
+    cm_idx         = ucp_ep_ext_control(ep)->cm_idx;
+    if (worker->cms[cm_idx].attr.max_conn_priv < sa_data_length) {
+        log_level = can_fallback ?  UCS_LOG_LEVEL_DIAG : UCS_LOG_LEVEL_ERROR;
         ucs_log(log_level,
                 "CM private data buffer is too small to pack UCP endpoint"
-                " info, ep %p service data %lu, address length %lu, cm %p"
-                " max_conn_priv %lu", ep, sizeof(ucp_wireup_sockaddr_data_t),
-                ucp_addr_size, worker->cms[cm_idx].cm,
-                worker->cms[cm_idx].attr.max_conn_priv);
+                " info, ep %p service data version %u, size %lu, address "
+                "length %lu, cm %p max_conn_priv %lu", ep, sa_data_version,
+                sa_data_length - ucp_addr_size, ucp_addr_size,
+                worker->cms[cm_idx].cm, worker->cms[cm_idx].attr.max_conn_priv);
         status = UCS_ERR_BUFFER_TOO_SMALL;
         goto err;
     }
 
-    sa_data = ucs_malloc(ucp_cm_priv_data_length(ucp_addr_size),
-                         "client_priv_data");
+    sa_data = ucs_malloc(sa_data_length, "client_priv_data");
     if (sa_data == NULL) {
         status = UCS_ERR_NO_MEMORY;
         goto err;
     }
 
-    sa_data->ep_id     = ucp_ep_local_id(ep);
-    sa_data->err_mode  = ucp_ep_config(ep)->key.err_mode;
-    sa_data->addr_mode = UCP_WIREUP_SA_DATA_CM_ADDR;
-    sa_data->dev_index = dev_index;
-    memcpy(sa_data + 1, ucp_addr, ucp_addr_size);
+    sa_data->ep_id = ucp_ep_local_id(ep);
+
+    if (sa_data_version == UCP_SA_DATA_VERSION_V1) {
+        sa_data->header       = ucp_ep_config(ep)->key.err_mode;
+        sa_data_v1            = ucs_derived_of(sa_data,
+                                               ucp_wireup_sockaddr_data_v1_t);
+        sa_data_v1->addr_mode = UCP_WIREUP_SA_DATA_CM_ADDR;
+        sa_data_v1->dev_index = 0;
+        worker_addr_p         = sa_data_v1 + 1;
+    } else {
+        ucs_assertv_always(sa_data_version == UCP_SA_DATA_VERSION_V2,
+                           "sa_data version: %u", sa_data_version);
+        sa_data->header = UCP_SA_DATA_VERSION_V2 <<
+                          UCP_SA_DATA_HEADER_VERSION_SHIFT;
+        worker_addr_p   = sa_data + 1;
+        if (ucp_ep_config(ep)->key.err_mode == UCP_ERR_HANDLING_MODE_PEER) {
+            sa_data->header |= UCP_SA_DATA_EXT_FLAG_ERR_PEER;
+        }
+    }
```

Comment:
pls add helper function(s)

Reply:
why? this code is present in one place only and the existing func does not look too big

Reply:
IMHO `ucp_cm_ep_priv_data_pack` looks pretty big after this PR

Reply:
ok, added func